### ISSUE TITLE
gcc error when installing with pip install

### ISSUE THREAD

**Issue by chananshgong** [hmmlearn/hmmlearn#205]
I get  `hmmlearn/_hmmc.c:239:28: fatal error: numpy/npy_math.h: No such file or directory` yet the installation seems to finish successfully.

requirements.txt file:
```
click==6.7
cython==0.25.2
joblib==0.11
numpy==1.12.1
pandas==0.19.2
python-speech-features==0.5
scikit-learn==0.18.1
scipy==0.19.0
hmmlearn==0.2.0
```

```  
Running setup.py bdist_wheel for hmmlearn: started
  Running setup.py bdist_wheel for hmmlearn: finished with status 'error'
  Complete output from command /opt/conda/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-8l6nu2n1/hmmlearn/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/tmpi_45qjtvpip-wheel- --python-tag cp36:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.6
  creating build/lib.linux-x86_64-3.6/hmmlearn
  copying hmmlearn/hmm.py -> build/lib.linux-x86_64-3.6/hmmlearn
  copying hmmlearn/utils.py -> build/lib.linux-x86_64-3.6/hmmlearn
  copying hmmlearn/base.py -> build/lib.linux-x86_64-3.6/hmmlearn
  copying hmmlearn/__init__.py -> build/lib.linux-x86_64-3.6/hmmlearn
  creating build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/test_utils.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/test_gaussian_hmm.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/test_gmm_hmm.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/test_multinomial_hmm.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/test_base.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  copying hmmlearn/tests/__init__.py -> build/lib.linux-x86_64-3.6/hmmlearn/tests
  running build_ext
  building 'hmmlearn._hmmc' extension
  creating build/temp.linux-x86_64-3.6
  creating build/temp.linux-x86_64-3.6/hmmlearn
  gcc -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/include/python3.6m -c hmmlearn/_hmmc.c -o build/temp.linux-x86_64-3.6/hmmlearn/_hmmc.o -O3
  hmmlearn/_hmmc.c:239:28: fatal error: numpy/npy_math.h: No such file or directory
   #include "numpy/npy_math.h"
                              ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for hmmlearn
  Running setup.py clean for hmmlearn
Successfully built python-speech-features
Failed to build hmmlearn
Installing collected packages: click, cython, joblib, numpy, pytz, python-dateutil, pandas, python-speech-features, scikit-learn, scipy, hmmlearn
  Running setup.py install for hmmlearn: started
    Running setup.py install for hmmlearn: finished with status 'done'
**Successfully installed** click-6.7 cython-0.25.2 **hmmlearn-0.2.0** joblib-0.11 numpy-1.12.1 pandas-0.19.2 python-dateutil-2.6.0 python-speech-features-0.5 pytz-2017.2 scikit-learn-0.18.1 scipy-0.19.0
```

**Comment by superbobry** [comment]
Could you install NumPy and retry the installation?

**Comment by chanansh** [comment]
Pip requirement file already has numpy

**Comment by chanansh** [comment]
I will try this but I am not sure it will help as I am building a Docker
and need one requirement file.

**Comment by solensolen** [comment]
I have the same problem.
try:apt-get install gcc
It works on my system.

**Comment by rrussell39** [comment]
apt-get install gcc 
returns:
gcc: error trying to exec 'cc1plus': execvp: No such file or directory
A subsequent call to apt-get install cc1plus shows no such package??

**Comment by StanSilas** [comment]
Any luck? I seem to be getting the same issue.

**Comment by robertocarta** [comment]
Any luck? Same problem here

**Comment by anntzer** [comment]
How was numpy installed?

**Comment by alegonz** [comment]
I just stumbled upon the same issue. Apparently it happens if you have `numpy` and `hmmlearn` together in the same requirements file. If you install them in separate `pip install` commands, the error goes away.

Should this issue be reopened?

**Comment by anntzer** [comment]
I can't repro in a clean venv:
```
$ python -mvenv /tmp/testenv
$ source /tmp/testenv/bin/activate
$ printf 'numpy\nhmmlearn\n' > requirements.txt
$ pip install -r requirements.txt
Collecting numpy (from -r requirements.txt (line 1))
  Using cached https://files.pythonhosted.org/packages/38/39/f73e104d44f19a6203e786d5204532e214443ea2954917b27f3229e7639b/numpy-1.15.4-cp37-cp37m-manylinux1_x86_64.whl
Collecting hmmlearn (from -r requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/94/1d/af4071cbe561f62d744946211414beb22ea7c72d4bb2f42636b6820b5ae5/hmmlearn-0.2.1.tar.gz (150kB)
    100% |████████████████████████████████| 153kB 377kB/s 
Collecting scikit-learn>=0.16 (from hmmlearn->-r requirements.txt (line 2))
  Using cached https://files.pythonhosted.org/packages/56/7d/0737aed3e157fb90a1eaecb1cbfa5742fa4206fb305c8f157b666b71da14/scikit_learn-0.20.0-cp37-cp37m-manylinux1_x86_64.whl
Collecting scipy>=0.13.3 (from scikit-learn>=0.16->hmmlearn->-r requirements.txt (line 2))
  Using cached https://files.pythonhosted.org/packages/40/de/0c22c6754370ba6b1fa8e53bd6e514d4a41a181125d405a501c215cbdbd6/scipy-1.1.0-cp37-cp37m-manylinux1_x86_64.whl
Installing collected packages: numpy, scipy, scikit-learn, hmmlearn
  Running setup.py install for hmmlearn ... done
Successfully installed hmmlearn-0.2.1 numpy-1.15.4 scikit-learn-0.20.0 scipy-1.1.0
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by alegonz** [comment]
Forgot to mention the environment I'm using. Apologies. I'm using Python 3.7 on docker. Perhaps it is related to the python version. Which version are you using?

This is my output log:

When done together:

`
$ docker run -it python:3.7.0-slim-stretch /bin/bash
root@464bff99684a:/# apt update && apt install -y gcc
...[log omitted for brevity]...
root@464bff99684a:/# printf 'numpy\nhmmlearn\n' > requirements.txt
root@464bff99684a:/# pip install -r requirements.txt
Collecting numpy (from -r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/38/39/f73e104d44f19a6203e786d5204532e214443ea2954917b27f3229e7639b/numpy-1.15.4-cp37-cp37m-manylinux1_x86_64.whl (13.8MB)
    100% |████████████████████████████████| 13.9MB 3.8MB/s 
Collecting hmmlearn (from -r requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/94/1d/af4071cbe561f62d744946211414beb22ea7c72d4bb2f42636b6820b5ae5/hmmlearn-0.2.1.tar.gz (150kB)
    100% |████████████████████████████████| 153kB 10.5MB/s 
Collecting scikit-learn>=0.16 (from hmmlearn->-r requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/56/7d/0737aed3e157fb90a1eaecb1cbfa5742fa4206fb305c8f157b666b71da14/scikit_learn-0.20.0-cp37-cp37m-manylinux1_x86_64.whl (5.3MB)
    100% |████████████████████████████████| 5.3MB 8.1MB/s 
Collecting scipy>=0.13.3 (from scikit-learn>=0.16->hmmlearn->-r requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/40/de/0c22c6754370ba6b1fa8e53bd6e514d4a41a181125d405a501c215cbdbd6/scipy-1.1.0-cp37-cp37m-manylinux1_x86_64.whl (31.2MB)
    100% |████████████████████████████████| 31.2MB 1.6MB/s 
Building wheels for collected packages: hmmlearn
  Running setup.py bdist_wheel for hmmlearn ... error
  Complete output from command /usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-j8mjhhm1/hmmlearn/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().rep
lace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-a1uz8zyn --python-tag cp37:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/_utils.py -> build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/utils.py -> build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/stats.py -> build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/__init__.py -> build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/hmm.py -> build/lib.linux-x86_64-3.7/hmmlearn
  copying hmmlearn/base.py -> build/lib.linux-x86_64-3.7/hmmlearn
  creating build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/conftest.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_base.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_multinomial_hmm.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/__init__.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_gmm_hmm.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_utils.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_gaussian_hmm.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  copying hmmlearn/tests/test_gmm_hmm_new.py -> build/lib.linux-x86_64-3.7/hmmlearn/tests
  running build_ext
  building 'hmmlearn._hmmc' extension
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/hmmlearn
  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include/python3.7m -c hmmlearn/_hmmc.c -o build/temp.linux-x86_64-3.7/hmmlearn/_hmmc.o -O3
  hmmlearn/_hmmc.c:565:28: fatal error: numpy/npy_math.h: No such file or directory
   #include "numpy/npy_math.h"
                              ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for hmmlearn
  Running setup.py clean for hmmlearn
Failed to build hmmlearn
Installing collected packages: numpy, scipy, scikit-learn, hmmlearn
  Running setup.py install for hmmlearn ... done
Successfully installed hmmlearn-0.2.1 numpy-1.15.4 scikit-learn-0.20.0 scipy-1.1.0
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
`

When done separate:
`
$ docker run -it python:3.7.0-slim-stretch /bin/bash
root@464bff99684a:/# apt update && apt install -y gcc
...[log omitted for brevity]...
root@1008e0a0355b:/# pip install numpy
Collecting numpy
  Downloading https://files.pythonhosted.org/packages/38/39/f73e104d44f19a6203e786d5204532e214443ea2954917b27f3229e7639b/numpy-1.15.4-cp37-cp37m-manylinux1_x86_64.whl (13.8MB)
    100% |████████████████████████████████| 13.9MB 3.6MB/s 
Installing collected packages: numpy
Successfully installed numpy-1.15.4
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
root@1008e0a0355b:/# pip install hmmlearn
Collecting hmmlearn
  Downloading https://files.pythonhosted.org/packages/94/1d/af4071cbe561f62d744946211414beb22ea7c72d4bb2f42636b6820b5ae5/hmmlearn-0.2.1.tar.gz (150kB)
    100% |████████████████████████████████| 153kB 9.3MB/s 
Requirement already satisfied: numpy in /usr/local/lib/python3.7/site-packages (from hmmlearn) (1.15.4)
Collecting scikit-learn>=0.16 (from hmmlearn)
  Downloading https://files.pythonhosted.org/packages/56/7d/0737aed3e157fb90a1eaecb1cbfa5742fa4206fb305c8f157b666b71da14/scikit_learn-0.20.0-cp37-cp37m-manylinux1_x86_64.whl (5.3MB)
    100% |████████████████████████████████| 5.3MB 10.1MB/s 
Collecting scipy>=0.13.3 (from scikit-learn>=0.16->hmmlearn)
  Downloading https://files.pythonhosted.org/packages/40/de/0c22c6754370ba6b1fa8e53bd6e514d4a41a181125d405a501c215cbdbd6/scipy-1.1.0-cp37-cp37m-manylinux1_x86_64.whl (31.2MB)
    100% |████████████████████████████████| 31.2MB 401kB/s 
Building wheels for collected packages: hmmlearn
  Running setup.py bdist_wheel for hmmlearn ... done
  Stored in directory: /root/.cache/pip/wheels/9a/a4/ee/917f0de81626b684fd2139ef5df47744c35ebeacc9e950487b
Successfully built hmmlearn
Installing collected packages: scipy, scikit-learn, hmmlearn
Successfully installed hmmlearn-0.2.1 scikit-learn-0.20.0 scipy-1.1.0
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
`

**Comment by qiwihui** [comment]
@alegonz In a clean env, running `pip install hmmlearn` will also repro this issue.

My env: docker `python:2.7.15-wheezy`

```sh
$ docker run -it python:2.7.15-wheezy bash
root@898ce6384731:/# 
root@898ce6384731:/# pip install hmmlearn
Collecting hmmlearn
  Downloading https://files.pythonhosted.org/packages/94/1d/af4071cbe561f62d744946211414beb22ea7c72d4bb2f42636b6820b5ae5/hmmlearn-0.2.1.tar.gz (150kB)
    100% |████████████████████████████████| 153kB 124kB/s 
Collecting numpy (from hmmlearn)
  Downloading https://files.pythonhosted.org/packages/de/37/fe7db552f4507f379d81dcb78e58e05030a8941757b1f664517d581b5553/numpy-1.15.4-cp27-cp27mu-manylinux1_x86_64.whl (13.8MB)
    100% |████████████████████████████████| 13.8MB 378kB/s 
Collecting scikit-learn>=0.16 (from hmmlearn)
  Downloading https://files.pythonhosted.org/packages/d4/3f/d9225d7d651a0e5562c0896bd9334fff40be1916bdfa07b150d0c0a7c7e2/scikit_learn-0.20.0-cp27-cp27mu-manylinux1_x86_64.whl (5.5MB)
    100% |████████████████████████████████| 5.5MB 502kB/s 
Collecting scipy>=0.13.3 (from scikit-learn>=0.16->hmmlearn)
  Downloading https://files.pythonhosted.org/packages/2a/f3/de9c1bd16311982711209edaa8c6caa962db30ebb6a8cc6f1dcd2d3ef616/scipy-1.1.0-cp27-cp27mu-manylinux1_x86_64.whl (30.8MB)
    100% |████████████████████████████████| 30.8MB 136kB/s 
Building wheels for collected packages: hmmlearn
  Running setup.py bdist_wheel for hmmlearn ... error
  Complete output from command /usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-iuIQ4c/hmmlearn/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-RRysv_ --python-tag cp27:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-2.7
  creating build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/hmm.py -> build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/base.py -> build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/utils.py -> build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/stats.py -> build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/_utils.py -> build/lib.linux-x86_64-2.7/hmmlearn
  copying hmmlearn/__init__.py -> build/lib.linux-x86_64-2.7/hmmlearn
  creating build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/conftest.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_gmm_hmm_new.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_gmm_hmm.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_gaussian_hmm.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_multinomial_hmm.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_utils.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/__init__.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  copying hmmlearn/tests/test_base.py -> build/lib.linux-x86_64-2.7/hmmlearn/tests
  running build_ext
  building 'hmmlearn._hmmc' extension
  creating build/temp.linux-x86_64-2.7
  creating build/temp.linux-x86_64-2.7/hmmlearn
  gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python2.7 -c hmmlearn/_hmmc.c -o build/temp.linux-x86_64-2.7/hmmlearn/_hmmc.o -O3
  hmmlearn/_hmmc.c:565:28: fatal error: numpy/npy_math.h: No such file or directory
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for hmmlearn
  Running setup.py clean for hmmlearn
Failed to build hmmlearn
Installing collected packages: numpy, scipy, scikit-learn, hmmlearn
  Running setup.py install for hmmlearn ... done
Successfully installed hmmlearn-0.2.1 numpy-1.15.4 scikit-learn-0.20.0 scipy-1.1.0
```

**Comment by anntzer** [comment]
But in both cases hmmlearn is actually correctly installed (and compiled): it does import just fine (and the exit code is 0).
Not sure about the error message but I guess it should go to the pip tracker?

**Comment by alegonz** [comment]
Sorry for the delayed reply.

@qiwihui 
I see, thanks for checking on py2. So it is unrelated to the python version.

@anntzer 
Yeah, indeed it can be imported and seems to work in both cases, and looks like a pip-related issue. Since we can still import it's ok, I guess, but this might bite back in the future. I'll try searching in the pip issue tracker if anything related comes up.

**Comment by harishkashyap** [comment]
Same issue with latest python.

**Comment by adam-hanna** [comment]
I'm also getting this exact same error: `hmmlearn/_hmmc.c:565:28: fatal error: numpy/npy_math.h: No such file or directory` however mine is ending with a status of 1 and not 0, and therefore I cannot import and use hmmlearn in my project.

**Comment by adam-hanna** [comment]
My issue was resolved by not pip installing with the `--user` flag for both numpy and hmmlearn. Dunno why that would have an effect.

**Comment by h12w** [comment]
This issue can be reproduced easily with:
`docker run --rm python:3.7.4 pip install hmmlearn==0.2.1`

However, it has been fixed in 0.2.2:
`docker run --rm python:3.7.4 pip install hmmlearn==0.2.2`

**Comment by czflshzy** [comment]
I met the same problem.Can someone help me?
